pandas/core/frame.py

```diff
@@ -4134,7 +4134,11 @@ def query(self, expr: str, inplace: bool = False, **kwargs):
         kwargs["level"] = kwargs.pop("level", 0) + 1
         kwargs["target"] = None
         res = self.eval(expr, **kwargs)
-
+        if not is_bool_dtype(res):
+            # Special condition to check when dealing with higher dimensions
+            if not (res.ndim > 1 and (res.dtypes == bool).all()):
```

Comment:
It seems you don't want to be passing `res` to `is_bool_dtype` if the ndim is 2, is that right? Can you rewrite to make this more clear, something like:

```
if res.ndim == 1:
    is_bool_result = is_bool_dtype(res)
else:
    is_bool_result = res.ndim > 1 and (res.dtypes == bool).all()
```

For the last line, I think you need to also be using `is_bool_dtype`, otherwise it won't recognize EAs.

Reply:
Nice ! Done